### PR TITLE
[SPARK-6152] Use shaded ASM5 to support closure cleaning of Java 8 compiled classes

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch modifies Spark's closure cleaner (and a few other places) to use ASM 5, which is necessary in order to support cleaning of closures that were compiled by Java 8.

In order to avoid ASM dependency conflicts, Spark excludes ASM from all of its dependencies and uses a shaded version of ASM 4 that comes from `reflectasm` (see [SPARK-782](https://issues.apache.org/jira/browse/SPARK-782) and #232). This patch updates Spark to use a shaded version of ASM 5.0.4 that was published by the Apache XBean project; the POM used to create the shaded artifact can be found at https://github.com/apache/geronimo-xbean/blob/xbean-4.4/xbean-asm5-shaded/pom.xml.

http://movingfulcrum.tumblr.com/post/80826553604/asm-framework-50-the-missing-migration-guide was a useful resource while upgrading the code to use the new ASM5 opcodes.

I also added a new regression tests in the `java8-tests` subproject; the existing tests were insufficient to catch this bug, which only affected Scala 2.11 user code which was compiled targeting Java 8.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Apparently ASM is a common source of Java 8 upgrade pain: http://product.hubspot.com/blog/upgrading-to-java-8-at-scale

**Comment by JoshRosen** [comment]
Jenkins retest this please.

**Comment by JoshRosen** [comment]
I suppose it would be nice to actually run the Java 8 tests to see whether they would have caught this.

**Comment by pwendell** [comment]
This looks fine (i.e. LGTM). However, we could also look into actually shading asm ourselves in our published artifacts, similar to how we now shade jetty and other things. I'm fine to just use this already shaded one too though. Seems to do no harm and will allow us to work with Java 8.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
I added a regression test to the `java8-tests` subproject and verified that it fails before this patch's changes.

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -560,6 +570,14 @@ object Unidoc {
   )
 }
 
+object Java8TestSettings {
+  import BuildCommons._
+
+  lazy val settings = Seq(
+    compilerJVMVersion := "1.8"
```
Comment (JoshRosen) [inline project/SparkBuild.scala]: Subtlety: this setting will break the `java8` tests when running on Scala 2.10.x, which does not support the emission of Java 8 bytecode. For this reason, we need separate settings for configuring the scalc and javac target JVM versions and need to only set scalac to 1.8 when running Scala 2.11.4 or higher.

**Comment by JoshRosen** [comment]
@marmbrus, it turns out that our java8-tests Maven profile disables the Scala compiler plugin (presumably because it was being used to compile Java sources and was hard to configure properly). I _could_ spend some time to make it so that the regression test for this patch can run in Maven, but I'd rather work on other things given how much time I've already sunk into this. Therefore, I just amended the "Building Spark" guide to contain the one-liner for running this in SBT. I'll defer proper configuration / execution of these tests to a followup / someone else.

**Comment by marmbrus** [comment]
I think the most important thing is that it will be run by the PR builder.

**Comment by JoshRosen** [comment]
The Java 8 tests currently aren't run at all (in fact, some of the streaming ones appear to be broken). In order to test this in the PR builder, I'll have to configure it to use JDK 8 and will have to supply the right profiles. I could probably auto-set the right profile based on the detected JDK version.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 2]
```diff
@@ -542,15 +543,7 @@ private object BoxingFinder {
...
-    // ASM4 doesn't support Java 8 classes, which requires ASM5.
-    // So if the class is ASM5 (E.g., java.lang.Long when using JDK8 runtime to run these codes),
-    // then ClassReader will throw IllegalArgumentException,
-    // However, since this is only for testing, it's safe to skip these classes.
-    try {
-      Some(new ClassReader(new ByteArrayInputStream(baos.toByteArray)))
-    } catch {
-      case _: IllegalArgumentException => None
-    }
+    Some(new ClassReader(new ByteArrayInputStream(baos.toByteArray)))
```
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Just realized that this no longer needs to return an `Option`; will update now to keep the code clean.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by marmbrus** [comment]
Thanks, I'm going to merge this to master and 1.6.  We should look at turning Java 8 test on in Jenkins as a follow up.

**Pull request closed without merging**